RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc

```diff
@@ -124,246 +157,804 @@ void L2MuonSeedGeneratorFromL1T::produce(edm::Event& iEvent, const edm::EventSet
-                output->push_back(L2MuonTrajectorySeed(seedTSOS,container,alongMomentum,
-                                  MuonRef(muColl,  distance(muColl->begin(muColl->getFirstBX()),it)  )));
+                else {
+                  if(useUnassociatedL1) {
+                    // convert the TSOS into a PTSOD
+                    PTrajectoryStateOnDet const & seedTSOS = trajectoryStateTransform::persistentState( newTSOS,newTSOSDet->geographicalId().rawId());
+                    output->push_back(L2MuonTrajectorySeed(seedTSOS,container,alongMomentum,
+                                      MuonRef(muColl,  distance(muColl->begin(muColl->getFirstBX()),it)  )));
+                  }
+                } 
               }
+            } 
+          }
+          else {
+            // convert the TSOS into a PTSOD
+            PTrajectoryStateOnDet const & seedTSOS = trajectoryStateTransform::persistentState( tsos, theid.rawId());
+            output->push_back(L2MuonTrajectorySeed(seedTSOS,container,alongMomentum,
+                              MuonRef(muColl,  distance(muColl->begin(muColl->getFirstBX()),it)  )));
+          }      
+        }  
+      }
+    }
+
+  } // End of matchType 0
+
+  //--- matchType > 0 : Updated logics
+  else if(matchType > 0) {
+
+    unsigned int NmuColl = muColl->size();
+
+    vector< vector<double> > dRmtx;
+    vector< vector<const TrajectorySeed *> > selOffseeds;
+
+    edm::Handle<edm::View<TrajectorySeed> > offlineSeedHandle;
+    if(useOfflineSeed) {
+      iEvent.getByToken(offlineSeedToken_, offlineSeedHandle);
+      unsigned int NofflineSeed = offlineSeedHandle->size();
+      LogTrace(metname) << "Number of offline seeds " << NofflineSeed << endl;
+
+      // Initialize dRmtx and selOffseeds
+      dRmtx = vector< vector<double> >(NmuColl, vector<double>(NofflineSeed, 999.0));
+      selOffseeds = vector< vector <const TrajectorySeed *> >(NmuColl, vector<const TrajectorySeed *>(NofflineSeed, nullptr));
+    }
+
+    // At least one L1 muons are associated with offSeed
+    bool isAsso = false;
+
+    //--- Fill dR matrix
+    for (int ibx = muColl->getFirstBX(); ibx <= muColl->getLastBX(); ++ibx) {
+      if (centralBxOnly_ && (ibx != 0)) continue;
+      for (auto it = muColl->begin(ibx); it != muColl->end(ibx); it++){
+
+        //Position of given L1mu
+        unsigned int imu = distance(muColl->begin(muColl->getFirstBX()),it);
+
+        unsigned int quality = it->hwQual();
+        int valid_charge = it->hwChargeValid();
+
+        float pt    =  it->pt();
+        float eta   =  it->eta();
+        float theta =  2*atan(exp(-eta));
+        float phi   =  it->phi();
+        int charge  =  it->charge();
+        // Set charge=0 for the time being if the valid charge bit is zero
+        if (!valid_charge) charge = 0;
+
+        int link = 36 + (int)(it -> tfMuonIndex() / 3.);
+        bool barrel = true;
+        if ( (link >= 36 && link <= 41) || (link >= 66 && link <= 71)) barrel = false;
```

Comment:
What are these magic numbers "36", "41", "71"?
Can they be retrieved from somewhere?